### PR TITLE
Remove _workQueue of webSocket set to NULL which may leads to crash

### LINKED ISSUE #21086: com.squareup.SocketRocket.NetworkThread(18): EXC_BAD_ACCESS
- [x] Review the documentation: https://facebook.github.io/react-native
  - [x] Search for existing issues: https://github.com/facebook/react-native/issues
  - [x] Use the latest React Native release: https://github.com/facebook/react-native/releases

## Environment
  React Native Environment Info:
    System:
      OS: macOS High Sierra 10.13.6
      CPU: x64 Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
      Memory: 611.18 MB / 8.00 GB
      Shell: 5.3 - /bin/zsh
    Binaries:
      Node: 9.5.0 - /usr/local/bin/node
      Yarn: 1.6.0 - /usr/local/bin/yarn
      npm: 6.2.0 - /usr/local/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    SDKs:
      iOS SDK:
        Platforms: iOS 11.4, macOS 10.13, tvOS 11.4, watchOS 4.3
    IDEs:
      Xcode: 9.4.1/9F2000 - /usr/bin/xcodebuild
    npmPackages:
      react: 16.4.1 => 16.4.1
      react-native: 0.56.0 => 0.56.0
    npmGlobalPackages:
      create-react-native-app: 1.0.0
      react-native-cli: 2.0.1

## Description
Crash randomly when run on `iOS`, the cause is when `WebSocket`'s delegate callback, the delegate already released, leads to `EXC_BAD_ACCESS`.

More detail can see #6117.

## Reproducible Demo
Crash randomly.

## potential fix `PR`
#19489

### PR DESCRIPTION
**Pull request by zhongwuzw**
Fixes #21086.

This PR fixes a crash caused by a race condition when `webSocket` deallocated and `NSStream` delegate callback, because `NSStream`'s delegate callback be called on `RCTSR_networkRunLoop`.

This PR mainly changes:

* Remove unnecessary `nil` operation in `dealloc` method.
* Add a new method `_scheduleCleanUp` to schedule `webSocket` cleanup also on `RCTSR_networkRunLoop`.
* In `stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode` delegate method, add a `wself` to make safe further.

## Test Plan

Reproduce steps:
Run any `react-native` project, and keep it running, it would crash at some point.

## Release Notes

[IOS] [BUGFIX] [RCTSRWebSocket] - No needs to set `NULL` explicitly.

### CODE REVIEW

**Comment by zhongwuzw** [comment]
Crash backtrace:

```
(lldb) bt
* thread #9, name = 'com.squareup.SocketRocket.NetworkThread', stop reason = EXC_BAD_ACCESS (code=1, address=0x20)
    frame #0: 0x0000000183a601a0 libobjc.A.dylib`objc_retain + 16
    frame #1: 0x0000000183a60218 libobjc.A.dylib`objc_storeStrong + 44
    frame #2: 0x000000010084a0c4 AwesomProject`__copy_helper_block_((null)=0x00000001c526d240, (null)=0x000000016fc91a08) at RCTSRWebSocket.m:1350
    frame #3: 0x000000018421c964 libsystem_blocks.dylib`_Block_copy + 140
    frame #4: 0x0000000100cc91b0 libdispatch.dylib`_dispatch_Block_copy + 64
    frame #5: 0x0000000100cca3f8 libdispatch.dylib`dispatch_async + 64
  * frame #6: 0x0000000100849538 AwesomProject`-[RCTSRWebSocket stream:handleEvent:](self=0x0000000101206cb0, _cmd="stream:handleEvent:", aStream=0x00000001c4105a90, eventCode=NSStreamEventErrorOccurred) at RCTSRWebSocket.m:1350
    frame #7: 0x00000001847c3b34 CoreFoundation`_signalEventSync + 212
    frame #8: 0x00000001847c3a40 CoreFoundation`_cfstream_shared_signalEventSync + 460
    frame #9: 0x0000000184833404 CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
    frame #10: 0x0000000184832c2c CoreFoundation`__CFRunLoopDoSources0 + 276
    frame #11: 0x000000018483079c CoreFoundation`__CFRunLoopRun + 1204
    frame #12: 0x0000000184750da8 CoreFoundation`CFRunLoopRunSpecific + 552
    frame #13: 0x00000001851c5674 Foundation`-[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 304
    frame #14: 0x000000010084b644 AwesomProject`-[_RCTSRRunLoopThread main](self=0x00000001c408f870, _cmd="main") at RCTSRWebSocket.m:1592
    frame #15: 0x00000001852d5efc Foundation`__NSThread__start__ + 1040
    frame #16: 0x00000001844b1220 libsystem_pthread.dylib`_pthread_body + 272
    frame #17: 0x00000001844b1110 libsystem_pthread.dylib`_pthread_start + 292
    frame #18: 0x00000001844afb10 libsystem_pthread.dylib`thread_start + 4
```

**Comment by zhongwuzw** [comment]
@hramos I sync the cleanup code of `RCTSRWebSocket` from [SRWebSocket](https://github.com/facebook/SocketRocket/blob/master/SocketRocket/SRWebSocket.m).

Enhance some cleanup logic of `*Stream`. 
The `PR` handles some edge cases, would you give me any thoughts about how to test?

**Comment by zhongwuzw** [comment]
@hramos Just ping. Many people face these crash things, please see #6117 .

**Comment by zhongwuzw** [comment]
@shergin , I found no people can handle this, but I see you work so much in `iOS`, please review it if you can. Thanks.

**Comment by zhongwuzw** [comment]
@facebook-github-bot marks #6117 resolved. But it's not.

**Comment by connoremma** [comment]
@zhongwuzw Still seeing #6117 on version: 

react-native-cli: 2.0.1
react-native: 0.55.4

**Comment by heyfrench** [comment]
@connoremma it's happened to me in production and debugging. I can post what I have from Crashlytics if that would help anybody.

FWIW, it hasn't happened since I implemented this patch a few weeks ago: https://github.com/facebook/react-native/issues/6117#issuecomment-405837832

**Comment by zhongwuzw** [comment]
@connoremma, this bug always exist, so no downgrading  we can have. 😂 Just like @heyfrench said, you can apply this patch using git, or just copy the files I changed in this `PR` in your project.

**Comment by ujwal-setlur** [comment]
I am running into this issue daily multiple times on my iOS simulator. I will try this patch

**Comment by ujwal-setlur** [comment]
@zhongwuzw Is there a bug open against react-native for this? If not, we should open one.

**Comment by zhongwuzw** [comment]
@ujwal-setlur There had one previous. Please see #6117 , but unfortunately @facebook-github-bot marks #6117 resolved.

**Comment by connoremma** [comment]
@zhongwuzw @heyfrench Thank you both for the recommendations, I have copied over the contents of Libraries/WebSocket/RCTSRWebSocket.m from this PR into my project, but unfortunately I am still having the error. Perhaps the script is necessary since I'm unsure if these files are created during `react-native-init` or `react-native run-ios`. My knowledge is limited on the topic so apologies for any misunderstanding

**Comment by pribeh** [comment]
This is an issue for us. Thanks for working on this @zhongwuzw. I believe the patch has resolved the issue for us but we're going to run more tests.

**Comment by mattijsf** [comment]
@zhongwuzw can you try rebasing? that might take care of the failing test cases as they are most likely unrelated to your changes.

**Comment by ujwal-setlur** [comment]
@zhongwuzw since I have applied this patch, I have not seen the problem. My app has been running for several hours. Thank you so much for this PR.

**Comment by zhongwuzw** [comment]
@pribeh @ujwal-setlur ❤️ 

@mattijsf Yeah, I can later if anyone can review this `PR` 😂 , maybe we need to wait a long time.

@connoremma You can try the git script to apply, if you face any error, please comment in this page, crash backtrace or something would be appreciated.

**Comment by ujwal-setlur** [comment]
I think it's best to open new bug report and link this PR to it. Maybe it will get attention then.

**Comment by hramos** [comment]
#6117 has been closed for a long time. Commenting on closed issues is rarely the best way to reach maintainers. I recommend you go ahead and file a new issue, even though this PR already exists - that way we can continue tracking the issue in case this PR does not get merged.

I'm going through PRs right now and it's hard for me, at a glance, to get an understanding of what you're trying to fix here and how your change fixed it. Perhaps you can elaborate a bit more on your PR description? You can edit the original post here on GitHub.

**Comment by zhongwuzw** [comment]
@hramos Thanks, I already file a new issue, please see #21086.

The reason for crash is the race condition when `webSocket` deallocated and `NSStream` delegate callback, because `NSStream`'s delegate callback be called on `RCTSR_networkRunLoop`.
So this `PR` mainly changes:
* Remove unnecessary `nil` operation in `dealloc` method.
* Add a new method `_scheduleCleanUp` to schedule `webSocket` cleanup also on `RCTSR_networkRunLoop`.
* In `stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode` delegate method, add a `wself` to make safe further.

File: Libraries/WebSocket/RCTSRWebSocket.m [hunk 1]
```diff
@@ -1315,6 +1310,8 @@ - (void)_sendFrameWithOpcode:(RCTSROpCode)opcode data:(id)data;
 
 - (void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode;
 {
+  __weak typeof(self) wself = self;
```
Comment (hramos) [inline Libraries/WebSocket/RCTSRWebSocket.m]: rename to weakSelf and move to line 1345

File: Libraries/WebSocket/RCTSRWebSocket.m [hunk 2]
```diff
@@ -1345,94 +1342,133 @@ - (void)stream:(NSStream *)aStream handleEvent:(NSStreamEvent)eventCode;
...
+  if (_cleanupScheduled) {
+    return;
+  }
+  
+  _cleanupScheduled = YES;
+  
+  // Cleanup NSStream delegate's in the same RunLoop used by the streams themselves:
+  // This way we'll prevent race conditions between handleEvent and SRWebsocket's dealloc
+  NSTimer *timer = [NSTimer timerWithTimeInterval:(0.0f) target:self selector:@selector(_cleanupSelfReference:) userInfo:nil repeats:NO];
+  [[NSRunLoop RCTSR_networkRunLoop] addTimer:timer forMode:NSDefaultRunLoopMode];
```
Comment (hramos) [inline Libraries/WebSocket/RCTSRWebSocket.m]: I'm a bit nervous about doing the NSTimer to the RCTSR_networkRunLoop and would prefer just deallocating it in advance. Is there no other way?

Comment (zhongwuzw) [reply Libraries/WebSocket/RCTSRWebSocket.m]: Yeah, I looked into this again, and we do all network callback on `_workQueue`, for the sake of thread safe, we put `_cleanupSelfReference` on `_workQueue` would be better.

**Comment by stephencookdev** [comment]
I can fairly reliably repro this issue by just running any react-native project in a Simulator, and leaving it. It's not easy to hit, but if you just leave your app running all day, you'll almost certainly see it's hit this error by the end of the day.

As for repro steps that are programatic, so we can write a test plan...  I've not been able to work out a way of forcing the race condition, short of just waiting 😞

Tbh I'm struggling to follow this PR fully... But am I right in understanding, @zhongwuzw, that specifically the issue is around `handleEvent` being executed _after_ the cleanup occurs? And if so, could we write an integration test (since, as I understand it, React iOS doesn't have unit tests?) that triggers this, and we can just test that no crash occurs?

**Comment by ishigamii** [comment]
Is this issue merged somewhere ? can we safely make the PR in our project to clear this issue that happens quite a lot ? 
Thanks a lot.

**Comment by ujwal-setlur** [comment]
It does not seem to be merged, but I have been applying this patch and have had no problems. Haven't seen this issue since.

**Comment by connoremma** [comment]
noob question; is RTCSRWebsocket.m generated during react-native init?

There have been times where I’ve replaced my copy with the one in this PR, and others where I’ve had to rebuild the project and started seeing it again, anyone familiar with this?

**Comment by tomaszwojcik** [comment]
Still struggling with the `com.squareup.SocketRocket.NetworkThread (9): signal SIGABRT` issue on `0.57.4` ☹️ Any ETA on the merge?

**Comment by zhongwuzw** [comment]
@tomaszwojcik Emm, seems we still need to wait to merge, but you can also just apply this patch first.

**Comment by AndrewJack** [comment]
FYI: We've applied this PR as a patch on top of 0.57 and we haven't seen the `com.squareup.SocketRocket.NetworkThread` crash since.

If there's any chance of this being rebased & merged, that would be awesome. 😁

**Comment by kelset** [comment]
Hey @zhongwuzw it seems that something went wrong during your rebase 😅

Probably it would be easier if you open a new PR instead? 🤔

**Comment by zhongwuzw** [comment]
@kelset Yes 😂 , I make a new `PR` #22439 . 

Closing in favor of #22439 .
